subtests/docker_cli/run_env/run_env.py

```diff
@@ -3,493 +3,123 @@
-            dc.clean_all(self.sub_stuff.get("containers", []))
-
-
-class port_base(run_env_base):
-
-    """
-    Base class for port test.
-    """
+    pass
+
+
+class spam(SubSubtest):
+
+    # Allows quickly rejecting container env. supplied vars
+    KEYPFX = '_x'
+    VALPFX = 'x_'
+
+    @classmethod
+    def make_test_dict(cls, howmany, keyiter=None, valiter=None):
+        xcept = ValueError("Must specify neither or both "
+                           "keyiter and valiter")
+        if ((keyiter is not None and valiter is None) or
+                (keyiter is None and valiter is not None)):
+            raise xcept
+        keyfmt = "%s%%x" % cls.KEYPFX
+        valfmt = "%s%%x" % cls.VALPFX
+        # Because of above, only need to check one
+        if keyiter is None:
+            keyiter = (keyfmt % x for x in xrange(0, howmany))
+            valiter = (valfmt % x for x in xrange(howmany, 0, -1))
+        return dict(zip(keyiter, valiter))
+
+    def save_testdata(self):
+        self.sub_stuff['output'] = os.path.join(self.tmpdir, 'env.output')
+        resultdir = self.parent_subtest.resultsdir
+        inputcopy = os.path.join(resultdir, "env.input")
+        with open(inputcopy, 'wb') as inpf:
+            for key, value in self.sub_stuff['envd'].iteritems():
+                inpf.write("%s=%s\n" % (key, value))
+
+    def initialize_cmdline(self):
+        subargs = ['--rm', '-v', '%s:/x:Z' % self.tmpdir]
+        for key, value in self.sub_stuff['envd'].iteritems():
+            subargs.append("-e")
+            subargs.append("%s=%s" % (key, value))
+        subargs.append(DockerImages(self).default_image)
+        subargs.append("bash -c 'env > /x/env.output'")
+        _ = DockerCmd(self, 'run', subargs, verbose=False)
+        _.quiet = True
+        self.sub_stuff['dkrcmd'] = _
+
+    def retrieve_testdata(self):
+        self.failif(not os.path.isfile(self.sub_stuff['output']),
+                    "Could not find %s" % self.sub_stuff['output'])
+        resultdir = self.parent_subtest.resultsdir
+        outputcopy = os.path.join(resultdir,
+                                  os.path.basename(self.sub_stuff['output']))
+        shutil.copyfile(self.sub_stuff['output'], outputcopy)
+        output = {}
+        with open(self.sub_stuff['output'], "r") as outf:
+            for line in outf:
+                key, value = line.split('=', 2)
+                key = key.strip()
+                value = value.strip()
+                if (not key.startswith(self.KEYPFX) or
+                        not value.startswith(self.VALPFX)):
+                    self.logdebug("Ignoring %s", line)
+                    # These are not the droids we're looking for
+                    continue
+
+                output[key] = value
+        return output
+
+    def validate_testdata(self, expected, actual):
+        # More helpful to debugging if done this way
+        for key, value in actual.iteritems():
```

Comment:
Nope, you're correct.  I remember struggling with this while writing it.  It is checking that keys didn't get passed, and also that keys and values didn't get mangled in some way by docker on input or on dump into results file.  Perhaps the thing to do here is check it both ways, then there can be no doubt.